wolfcrypt/src/dsa.c

```diff
@@ -654,263 +654,471 @@ int wc_DsaExportKeyRaw(DsaKey* dsa, byte* x, word32* xSz, byte* y, word32* ySz)
+        goto out;
+    }
 
     /* kInv = 1/k.b mod q */
-    if (ret == 0 && mp_invmod(&k, &key->q, &kInv) != MP_OKAY)
+    if (mp_invmod(k, &key->q, kInv) != MP_OKAY) {
         ret = MP_INVMOD_E;
+        goto out;
+    }
 
-    if (ret == 0 && mp_mod(&r, &key->q, &r) != MP_OKAY)
+    if (mp_mod(r, &key->q, r) != MP_OKAY) {
         ret = MP_MOD_E;
+        goto out;
+    }
 
     /* s = x.r */
-    if (ret == 0 && mp_mul(&key->x, &r, &s) != MP_OKAY)
+    if (mp_mul(&key->x, r, s) != MP_OKAY) {
         ret = MP_MUL_E;
+        goto out;
+    }
 
     /* s = x.r/k.b */
-    if (ret == 0 && mp_mulmod(&s, &kInv, &key->q, &s) != MP_OKAY)
+    if (mp_mulmod(s, kInv, &key->q, s) != MP_OKAY) {
         ret = MP_MULMOD_E;
+        goto out;
+    }
 
     /* H = H/k.b */
-    if (ret == 0 && mp_mulmod(&H, &kInv, &key->q, &H) != MP_OKAY)
+    if (mp_mulmod(H, kInv, &key->q, H) != MP_OKAY) {
         ret = MP_MULMOD_E;
+        goto out;
+    }
 
-    /* s = H/k.b + x.r/k.b
-         = (H + x.r)/k.b */
-    if (ret == 0 && mp_add(&s, &H, &s) != MP_OKAY)
+    /* s = H/k.b + x.r/k.b = (H + x.r)/k.b */
+    if (mp_add(s, H, s) != MP_OKAY) {
         ret = MP_ADD_E;
+        goto out;
+    }
 
-    /* s = b.(e + x.r)/k.b
-         = (e + x.r)/k */
-    if (ret == 0 && mp_mulmod(&s, &b, &key->q, &s) != MP_OKAY)
+    /* s = b.(e + x.r)/k.b = (e + x.r)/k */
+    if (mp_mulmod(s, b, &key->q, s) != MP_OKAY) {
         ret = MP_MULMOD_E;
+        goto out;
+    }
 
     /* s = (e + x.r)/k */
-    if (ret == 0 && mp_mod(&s, &key->q, &s) != MP_OKAY)
+    if (mp_mod(s, &key->q, s) != MP_OKAY) {
         ret = MP_MOD_E;
+        goto out;
+    }
 #endif
 
     /* detect zero r or s */
-    if (ret == 0 && (mp_iszero(&r) == MP_YES || mp_iszero(&s) == MP_YES))
+    if ((mp_iszero(r) == MP_YES) || (mp_iszero(s) == MP_YES)) {
         ret = MP_ZERO_E;
+        goto out;
+    }
 
     /* write out */
-    if (ret == 0)  {
-        int rSz = mp_unsigned_bin_size(&r);
-        int sSz = mp_unsigned_bin_size(&s);
+    {
+        int rSz = mp_unsigned_bin_size(r);
+        int sSz = mp_unsigned_bin_size(s);
 
         while (rSz++ < DSA_HALF_SIZE) {
             *out++ = 0x00;  /* pad front with zeros */
         }
 
-        if (mp_to_unsigned_bin(&r, out) != MP_OKAY)
+        if (mp_to_unsigned_bin(r, out) != MP_OKAY)
             ret = MP_TO_E;
         else {
             out = tmp + DSA_HALF_SIZE;  /* advance to s in output */
             while (sSz++ < DSA_HALF_SIZE) {
                 *out++ = 0x00;  /* pad front with zeros */
             }
-            ret = mp_to_unsigned_bin(&s, out);
+            ret = mp_to_unsigned_bin(s, out);
         }
     }
 
-    ForceZero(buffer, sz);
-    mp_forcezero(&kInv);
-    mp_forcezero(&k);
-#ifndef WOLFSSL_MP_INVMOD_CONSTANT_TIME
-    mp_forcezero(&b);
+  out:
 
-    mp_clear(&b);
+#ifdef WOLFSSL_SMALL_STACK
+    if (k) {
+        if (ret != MP_INIT_E) {
+            mp_forcezero(k);
+            mp_clear(k);
```

Comment:
mp_forcezero() does the same as mp_clear() but ensures the zeroing happens.
i.e. don't need to call mp_clear()

Reply:
I suspected as much but aped the incumbent code out of expedience.  Thanks.  I fixed the incumbent code in `wc_DsaSign()` too, FWIW.